### PR TITLE
[CDAP-20386] Reset to remote branch on repeated git clone

### PR DESCRIPTION
**Pull request by arjan-bal**
[CDAP-20386](https://cdap.atlassian.net/browse/CDAP-20386)

On calling repeated calls to `RepositoryManager#clone()`, do a hard reset to remote branch.

## Tests
Two unit tests were already calling `manager.close()` followed by `manager.cloneRemote()` to get changes in the remote repo, the close call has been removed and the tests are passing.

[CDAP-20386]: https://cdap.atlassian.net/browse/CDAP-20386?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java [hunk 1]
```diff
@@ -181,7 +186,8 @@ public String getFileHash(Path relativePath, String commitHash) throws IOExcepti
       // Find the node representing the exact file path in the tree.
       TreeWalk walk = TreeWalk.forPath(git.getRepository(), realRelativePath.toString(), commit.getTree());
       if (walk == null) {
-        throw new NotFoundException("File not found in Git revision tree.");
+        throw new NotFoundException(
+          "File not found in Git revision tree. Try re-cloning remote if file was added recently.");
```
Comment (samdgupi) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: nit: This string should not get exposed to the user. End-User will not have any way to re-clone remote

File: cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/RepositoryManagerTest.java [hunk 2]
```diff
@@ -271,6 +269,18 @@ public void testGetFileHashSymlinks() throws Exception {
     manager.close();
```
Comment (de-lan) [inline cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/RepositoryManagerTest.java]: This one could also get removed?

Comment (arjan-bal) [reply cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/RepositoryManagerTest.java]: The final close is to clean up the resources used by the manager. close() calls in the middle were there to force a re-clone, they have been removed.

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java [hunk 3]
```diff
@@ -320,4 +340,25 @@ private void popStash(@Nullable RevCommit stashID) throws GitAPIException {
     LOG.warn("Couldn't find Git stash with ID {} to drop. Ignoring missing stash.", stashID.getName());
   }
 
+  /**
+   * Does a <a href="https://git-scm.com/docs/git-reset"> hard reset </a> to the remote branch being tracked.
+   */
+  private String resetToCleanBranch() throws IOException, GitAPIException {
```
Comment (samdgupi) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: should not it just be `hardReset()`? IIRC `git reset --hard` does exactly what we want to achieve

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java [hunk 4]
```diff
@@ -192,29 +198,30 @@ public String getFileHash(Path relativePath, String commitHash) throws IOExcepti
...
     RepositoryConfig repositoryConfig = sourceControlConfig.getRepositoryConfig();
-    credentialsProvider.refresh();
     CloneCommand command =
       createCommand(Git::cloneRepository).setURI(repositoryConfig.getLink()).setDirectory(getRepositoryRoot().toFile());
     String branch = getBranchRefName(repositoryConfig.getDefaultBranch());
     if (branch != null) {
       command.setBranchesToClone(Collections.singleton((branch))).setBranch(branch);
     }
     git = command.call();
+    remoteBranchName = git.getRepository().getBranch();
```
Comment (tivv) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: Why do you need this mutable state? You can always do `git.getRepository().getBranch()` to query it, can't you? Or alternatively you can use a fixed local branch name. Mutable state is something to avoid if possible

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java [hunk 5]
```diff
@@ -320,4 +340,25 @@ private void popStash(@Nullable RevCommit stashID) throws GitAPIException {
...
 
+  /**
+   * Does a <a href="https://git-scm.com/docs/git-reset"> hard reset </a> to the remote branch being tracked.
+   */
+  private String resetToCleanBranch() throws IOException, GitAPIException {
+    List<RemoteConfig> remoteConfigs = git.remoteList().call();
+    if (remoteConfigs.isEmpty()) {
+      throw new IllegalStateException("No remotes found for git repository.");
+    }
+    if (remoteConfigs.size() > 1) {
```
Comment (tivv) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: nit: you can do `!= 1` and use 1 check instead of 2. Anyway this is something that should never fail. Still good to validate though

Comment (arjan-bal) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: Merged both the checks.

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java (hunk 4 above)
Comment (arjan-bal) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: I did it initially since the call to getBranch can throw exceptions and it is not expected to change. I removed the private non-final variable now.

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java [hunk 6]
```diff
@@ -181,33 +184,34 @@ public String getFileHash(Path relativePath, String commitHash) throws IOExcepti
       // Find the node representing the exact file path in the tree.
       TreeWalk walk = TreeWalk.forPath(git.getRepository(), realRelativePath.toString(), commit.getTree());
       if (walk == null) {
-        throw new NotFoundException("File not found in Git revision tree.");
+        throw new NotFoundException(
+          "File not found in Git revision tree. Try re-cloning remote if file was added recently.");
```
Comment (tivv) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: This exception may pop up to user and user has no way to do this suggestion. Let's not add a suggestion user can't do.

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java [hunk 7]
```diff
@@ -320,4 +337,22 @@ private void popStash(@Nullable RevCommit stashID) throws GitAPIException {
...
+    List<RemoteConfig> remoteConfigs = git.remoteList().call();
+    if (remoteConfigs.size() != 1) {
+      throw new IllegalStateException(String.format("%s remotes found for git repository, expected 1.",
+                                                    remoteConfigs.size()));
+    }
+    RemoteConfig origin = remoteConfigs.get(0);
+    createCommand(git::fetch).setRefSpecs(origin.getFetchRefSpecs()).call();
+    git.reset()
+      .setMode(ResetCommand.ResetType.HARD)
+      .setRef(getRemoteBranchRefName(remoteBranchName, origin.getName()))
```
Comment (tivv) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: We also have a TODO to fetch at this point. Is it going to be in a different PR?

**Review by tivv — COMMENTED** [review]
Please add a JIRA to the PR title

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java (hunk 7 above)
Comment (arjan-bal) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: All the operations that are performed are on top top of the head of the default branch in the remote repo or the branch specified in the repository config stored for the CDAP namespace. Its not possible for someone to request a commit ID which isn't there in the repository manager unless either of the following are true:
1.  Commit is in the the future, i.e the remote head has moved forward and the caller got the commit ID from somewhere other than the repository manager.
2. Commit is in the history, i.e the remote history has changed (force push) so the commit ID is no longer a part of the remote branch that has been cloned.
3. Commit is in another branch.

To resolve 1, we can fetch and checkout the new head, but we have 2 problems:
*  Restoring the stashed changes can cause conflicts.
*  After going back to the previous commit once we have the file hash, we know that we're no longer at the head. The assumption of always being on the head is invalid. Future pushes will fail unless the caller does a cloneRemote() to get back to head and re-apply the changes.

To resolve 2, fetching the commit will help.
To resolve 3, the caller should create a new repository manager as the current implementation doesn't handle changing configs, eg: a push would result in pushing to the old branch.

Trying to handle 1 and 3 will make the implementation complex and will probably need the caller to do a re-clone / create a new manager later on anyway.
Situation 2 is unlikely, it would happen when the commit ID was stored somewhere and a re-clone was performed while the remote was force pushed. If the same repository manager provided the initial commit ID, it will still have the commit with it. If the commit ID is stored in persistent storage, callers can save the git hash along with it. 

Due to the additional edge cases, I think significantly its easier if the caller just re-clones and retries the entire pull/list apps flow again. Push will not face this problem since the new commit is made in the local repo.

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java (hunk 1 above)
Comment (arjan-bal) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: Removed.

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java (hunk 6 above)
Comment (arjan-bal) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: Removed.

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java (hunk 3 above)
Comment (arjan-bal) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: We need to fetch the latest heads also. This method is essentially a fetch followed by a rest hard. There a call to clean() so that any untracked and ignored files are also discarded and you get a clean slate.

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java (hunk 7 above)
Comment (tivv) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: Well, it's the requirement of CDAP-20386 to handle (1). It's explicitly stated in the JIRA

Comment (tivv) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: I also don't see what are the problems with doing a fetch and then checking out a proper commit in all cases. Given you know the commit you don't need to know if it's same branch, different branch or force pushed branch. You just reset the current local branch to the proper remote commit

Comment (tivv) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: Could you elaborate on the problems? 

I don't think we plan to use stash in the local repo, do we? If so, what for?

Also you can't assume you are always on the remote branch head as there is always some time between the fetch and the push and remote branch can change in between. Local repo state does not matter here.

Comment (tivv) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: And cloning won't fix it. I don't see how clone is different from fetch+checkout here

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java [hunk 8]
```diff
@@ -181,33 +184,33 @@ public String getFileHash(Path relativePath, String commitHash) throws IOExcepti
...
    * @return the commit ID of the present HEAD.
    * @throws GitAPIException               when a Git operation fails.
    * @throws IOException                   when file or network I/O fails.
    * @throws AuthenticationConfigException when there is a failure while fetching authentication credentials for Git.
    */
   public String cloneRemote() throws Exception {
+    credentialsProvider.refresh();
     if (git != null) {
-      return resolveHead().getName();
+      return resetToCleanBranch(git.getRepository().getBranch());
```
Comment (iam-divya) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java:287]: Will we have any check for local changes before reset and getting confirmation from user for losing local changes? There can be more than one user uploading pipelines to git isn't it?

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java [hunk 9]
```diff
@@ -181,33 +184,33 @@ public String getFileHash(Path relativePath, String commitHash) throws IOExcepti
       // Find the node representing the exact file path in the tree.
       TreeWalk walk = TreeWalk.forPath(git.getRepository(), realRelativePath.toString(), commit.getTree());
       if (walk == null) {
-        throw new NotFoundException("File not found in Git revision tree.");
+        throw new NotFoundException( "File not found in Git revision tree.");
```
Comment (albertshau) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: remove extra space

**Comment by tivv** [comment]
@arjan-bal could you please rebase and resolve conflicts?

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java (hunk 8 above)
Comment (arjan-bal) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java:287]: No, this method will force reset. An additional parameter can be passed to "force" a reset or raise an exception, but there is no use case for it. Presently a new repository manager is created for each push/pull/list operation, so the same repository manager can't have multiple users.
The java doc mentions that the repository manager is not thread safe. The users need to hold a lock on the repository manager till the operation is complete.

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java (hunk 7 above)
Comment (arjan-bal) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: The stashing and checking out of commits is removed. Earlier, to follow symlinks the method was checking out the commit. Now we aren't following symlinks. To restore the state before getting the file hash, the stashing was being done.

**Comment by arjan-bal** [comment]
> @arjan-bal could you please rebase and resolve conflicts?

@tivv Rebased and replied to previous comments.

**Review by albertshau — APPROVED** [review]
lgtm after checkstyle fixes

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java (hunk 9 above)
Comment (arjan-bal) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java]: Fixed.
